### ISSUE TITLE
Pulling test images not Sonobuoy from private registry with secret.

### ISSUE THREAD

**Issue by kobemtl** [vmware-tanzu/sonobuoy#1038]
I am able to use sonobuoy image from private registry by following [this document](https://sonobuoy.io/docs/v0.16.5/pullsecrets/). Thanks for that. However, for those e2e test images I got image pull error. Sonobuoy creates new namespace but there is no secret key and imagepullsecrets  created. Below is my config for the job. Can anybody help me on this? Appreciate it. 

```yaml
---
apiVersion: v1
kind: Namespace
metadata:
  name: sonobuoy
---
apiVersion: v1
data:
  .dockerconfigjson: JSONSECRET
kind: Secret
metadata:
  name: sonobuoydockerkey
  namespace: sonobuoy
type: kubernetes.io/dockerconfigjson
---
apiVersion: v1
kind: ServiceAccount
metadata:
  labels:
    component: sonobuoy
  name: sonobuoy-serviceaccount
  namespace: sonobuoy
imagePullSecrets:
- name: sonobuoydockerkey  
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  labels:
    component: sonobuoy
  name: sonobuoy-serviceaccount-sonobuoy
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: sonobuoy-serviceaccount
subjects:
- kind: ServiceAccount
  name: sonobuoy-serviceaccount
  namespace: sonobuoy
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  labels:
    component: sonobuoy
  name: sonobuoy-serviceaccount
rules:
- apiGroups:
  - '*'
  resources:
  - '*'
  verbs:
  - '*'
- nonResourceURLs:
  - '/metrics'
  - '/logs'
  - '/logs/*'
  verbs:
  - 'get'
---
apiVersion: v1
data:
  config.json: |
    {"Description":"DEFAULT","UUID":"51d479c2-6238-4484-ad9b-ac3e8bbc4251","Version":"v0.16.4","ResultsDir":"/tmp/sonobuoy","Resources":["apiservices","certificatesigningrequests","clusterrolebindings","clusterroles","componentstatuses","configmaps","controllerrevisions","cronjobs","customresourcedefinitions","daemonsets","deployments","endpoints","ingresses","jobs","leases","limitranges","mutatingwebhookconfigurations","namespaces","networkpolicies","nodes","persistentvolumeclaims","persistentvolumes","poddisruptionbudgets","pods","podlogs","podsecuritypolicies","podtemplates","priorityclasses","replicasets","replicationcontrollers","resourcequotas","rolebindings","roles","servergroups","serverversion","serviceaccounts","services","statefulsets","storageclasses","validatingwebhookconfigurations","volumeattachments"],"Filters":{"Namespaces":".*","LabelSelector":""},"Limits":{"PodLogs":{"Namespaces":"","SonobuoyNamespace":true,"FieldSelectors":[],"LabelSelector":"","Previous":false,"SinceSeconds":null,"SinceTime":null,"Timestamps":false,"TailLines":null,"LimitBytes":null,"LimitSize":"","LimitTime":""}},"QPS":30,"Burst":50,"Server":{"bindaddress":"0.0.0.0","bindport":8080,"advertiseaddress":"","timeoutseconds":10800},"Plugins":[{"name":"e2e"},{"name":"systemd-logs"}],"PluginSearchPath":["./plugins.d","/etc/sonobuoy/plugins.d","~/sonobuoy/plugins.d"],"Namespace":"sonobuoy","WorkerImage":"sonobuoy/sonobuoy:v0.16.4","ImagePullPolicy":"IfNotPresent","ImagePullSecrets":"sonobuoydockerkey","ProgressUpdatesPort":"8099"}
kind: ConfigMap
metadata:
  labels:
    component: sonobuoy
  name: sonobuoy-config-cm
  namespace: sonobuoy
---
apiVersion: v1
data:
  plugin-0.yaml: |
    extra-volumes:
    - configMap:
        name: repolist-cm
      name: repolist-vol
    sonobuoy-config:
      driver: Job
      plugin-name: e2e
      result-format: junit
    spec:
      command:
      - /run_e2e.sh
      env:
      - name: E2E_FOCUS
        value: \[Conformance\]
      - name: E2E_SKIP
        value: \[Disruptive\]|NoExecuteTaintManager
      - name: E2E_PARALLEL
        value: "1"
      - name: E2E_USE_GO_RUNNER
        value: "true"
      - name: KUBE_TEST_REPO_LIST
        value: /tmp/sonobuoy/repo-list.yaml
      image: DOCKERREGISTRY/google-containers/conformance:v1.14.8
      imagePullPolicy: Always
      name: e2e
      resources: {}
      volumeMounts:
      - mountPath: /tmp/results
        name: results
      - mountPath: /tmp/sonobuoy
        name: repolist-vol
  plugin-1.yaml: |
    sonobuoy-config:
      driver: DaemonSet
      plugin-name: systemd-logs
      result-format: raw
    spec:
      command:
      - /bin/sh
      - -c
      - /get_systemd_logs.sh && while true; do echo "Sleeping for 1h to avoid daemonset
        restart"; sleep 3600; done
      env:
      - name: CHROOT_DIR
        value: /node
      - name: RESULTS_DIR
        value: /tmp/results
      - name: NODE_NAME
        valueFrom:
          fieldRef:
            fieldPath: spec.nodeName
      image: DOCKERREGISTRY/heptio-images/sonobuoy-plugin-systemd-logs:latest
      imagePullPolicy: Always
      name: systemd-logs
      resources: {}
      securityContext:
        privileged: true
      volumeMounts:
      - mountPath: /tmp/results
        name: results
      - mountPath: /node
        name: root
kind: ConfigMap
metadata:
  labels:
    component: sonobuoy
  name: sonobuoy-plugins-cm
  namespace: sonobuoy
---
apiVersion: v1
kind: Pod
metadata:
  labels:
    component: sonobuoy
    run: sonobuoy-master
    tier: analysis
  name: sonobuoy
  namespace: sonobuoy
spec:
  containers:
  - command:
    - /bin/bash
    - -c
    - /sonobuoy master --no-exit=true -v 3 --logtostderr
    env:
    - name: SONOBUOY_ADVERTISE_IP
      valueFrom:
        fieldRef:
          fieldPath: status.podIP
    image: sonobuoy/sonobuoy:v0.16.4
    imagePullPolicy: IfNotPresent
    name: kube-sonobuoy
    volumeMounts:
    - mountPath: /etc/sonobuoy
      name: sonobuoy-config-volume
    - mountPath: /plugins.d
      name: sonobuoy-plugins-volume
    - mountPath: /tmp/sonobuoy
      name: output-volume
  imagePullSecrets:
  - name: sonobuoydockerkey
  restartPolicy: Never
  serviceAccountName: sonobuoy-serviceaccount
  tolerations:
  - key: "kubernetes.io/e2e-evict-taint-key"
    operator: "Exists"
  volumes:
  - configMap:
      name: sonobuoy-config-cm
    name: sonobuoy-config-volume
  - configMap:
      name: sonobuoy-plugins-cm
    name: sonobuoy-plugins-volume
  - emptyDir: {}
    name: output-volume
---
apiVersion: v1
data:
  repo-list.yaml: |
    dockerLibraryRegistry: DOCKERREGISTRY
    e2eRegistry: DOCKERREGISTRY/kubernetes-e2e-test-images
    gcRegistry: DOCKERREGISTRY
    etcdRegistry: DOCKERREGISTRY/coreos
    privateRegistry: DOCKERREGISTRY/k8s-authenticated-test
    sampleRegistry: DOCKERREGISTRY/google-samples
kind: ConfigMap
metadata:
  name: repolist-cm
  namespace: sonobuoy
---
apiVersion: v1
kind: Service
metadata:
  labels:
    component: sonobuoy
    run: sonobuoy-master
  name: sonobuoy-master
  namespace: sonobuoy
spec:
  ports:
  - port: 8080
    protocol: TCP
    targetPort: 8080
  selector:
    run: sonobuoy-master
  type: ClusterIP
```

**Comment by johnSchnake** [comment]
It sounds like you are interested in running tests using a completely separate registry than the defaults? Not just Sonobuoy, but also the e2e tests (and maybe systemd-logs as well), is that right?

If that is the case you need to also look at the docs about testing in [airgapped](sonobuoy.io/docs/airgap) environments.

Sonobuoy has some commands `sonobuoy images` to help with moving the images from the public locations to your own private one.

But as of now there are a few images you have to move yourself, like the e2e image. See #1028 for details. 

After taking some time to look at those docs, let me know if that answers your question or you need additional guidance?

**Comment by kobemtl** [comment]
Thank you so much for answering John.

I have been follow the airgapped documents. All necessary images have been pull/push to our private registry. But like I mentioned, when Sonobuoy creates new namespace the secret not been created as long as the imagepullsecrets. Therefore I got unauthorized error. Please take a look at the details show below. Let me know further information need it. 

### Docker image
```shell
$docker search $PRIVATE_REG/nginx:1.14-alpine  
                                                             
NAME                                                                      DESCRIPTION         STARS               OFFICIAL            AUTOMATED
$PRIVATE_REG/nginx:1.14-alpine                       0
```
### Cluster pods
```shell
$kubectl get po -A

NAMESPACE         NAME                                                      READY   STATUS             RESTARTS   AGE
daemonsets-6591   daemon-set-6bkqd                                          0/1     ImagePullBackOff   0          3m35s
daemonsets-6591   daemon-set-9sqpt                                          0/1     ImagePullBackOff   0          3m35s
kube-system       aws-node-42kkn                                            1/1     Running            0          116m
kube-system       aws-node-82g22                                            1/1     Running            0          117m
kube-system       cluster-autoscaler-57fd5dcfdc-mmz7b                       1/1     Running            0          119m
kube-system       coredns-78768fbc56-p7p9d                                  1/1     Running            0          127m
kube-system       coredns-78768fbc56-svj5p                                  1/1     Running            0          127m
kube-system       kube-proxy-6656d                                          1/1     Running            0          120m
kube-system       kube-proxy-pr7qp                                          1/1     Running            0          120m
sonobuoy          sonobuoy                                                  1/1     Running            0          3m39s
sonobuoy          sonobuoy-e2e-job-934d8d246b794e97                         2/2     Running            0          3m38s
sonobuoy          sonobuoy-systemd-logs-daemon-set-4f743dd305694dd9-57wn6   2/2     Running            0          3m38s
sonobuoy          sonobuoy-systemd-logs-daemon-set-4f743dd305694dd9-lstcc   2/2     Running            0          3m38s
```

### Pod errors
```shell
$kubectl -n  daemonsets-6591 describe po daemon-set-6bkqd 

Name:           daemon-set-6bkqd
Namespace:      daemonsets-6591
Priority:       0
Node:           ip-10-6-215-185.ca-central-1.compute.internal/10.6.215.185
Start Time:     Mon, 09 Dec 2019 14:54:14 -0500
Labels:         controller-revision-hash=78c64684c7
                daemonset-name=daemon-set
                pod-template-generation=1
Annotations:    kubernetes.io/psp: e2e-test-privileged-psp
Status:         Pending
IP:             10.6.215.245
IPs:            <none>
Controlled By:  DaemonSet/daemon-set
Containers:
  app:
    Container ID:   
    Image:          $PRIVATE_REG/nginx:1.14-alpine
    Image ID:       
    Port:           9376/TCP
    Host Port:      0/TCP
    State:          Waiting
      Reason:       ImagePullBackOff
    Ready:          False
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-phnph (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  default-token-phnph:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-phnph
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/disk-pressure:NoSchedule
                 node.kubernetes.io/memory-pressure:NoSchedule
                 node.kubernetes.io/not-ready:NoExecute
                 node.kubernetes.io/pid-pressure:NoSchedule
                 node.kubernetes.io/unreachable:NoExecute
                 node.kubernetes.io/unschedulable:NoSchedule
Events:
  Type     Reason     Age                   From                                                    Message
  ----     ------     ----                  ----                                                    -------
  Normal   Scheduled  4m2s                  default-scheduler                                       Successfully assigned daemonsets-6591/daemon-set-6bkqd to ip-10-6-215-185.ca-central-1.compute.internal
  Normal   Pulling    2m28s (x4 over 4m1s)  kubelet, ip-10-6-215-185.ca-central-1.compute.internal  Pulling image "$PRIVATE_REG/nginx:1.14-alpine"
  Warning  Failed     2m28s (x4 over 4m1s)  kubelet, ip-10-6-215-185.ca-central-1.compute.internal  Failed to pull image "$PRIVATE_REG/nginx:1.14-alpine": rpc error: code = Unknown desc = Error response from daemon: unauthorized: access to the requested resource is not authorized
  Warning  Failed     2m28s (x4 over 4m1s)  kubelet, ip-10-6-215-185.ca-central-1.compute.internal  Error: ErrImagePull
  Normal   BackOff    2m17s (x6 over 4m)    kubelet, ip-10-6-215-185.ca-central-1.compute.internal  Back-off pulling image "$PRIVATE_REG/nginx:1.14-alpine"
  Warning  Failed     2m4s (x7 over 4m)     kubelet, ip-10-6-215-185.ca-central-1.compute.internal  Error: ImagePullBackOff
```

**Comment by kobemtl** [comment]
Here is the different between namespace sonobuoy and daemonsets-6591

### sonobuoy
```bash
$ kubectl get serviceaccounts -n sonobuoy 
NAME                      SECRETS   AGE
default                   1         16m
sonobuoy-serviceaccount   1         16m

$ kubectl get secret -n sonobuoy 
NAME                                            TYPE                                  DATA   AGE
default-token-qg7sk                             kubernetes.io/service-account-token   3      16m
sonobuoy-plugin-e2e-934d8d246b794e97            kubernetes.io/tls                     2      16m
sonobuoy-plugin-systemd-logs-4f743dd305694dd9   kubernetes.io/tls                     2      16m
sonobuoy-serviceaccount-token-rkgh6             kubernetes.io/service-account-token   3      16m
sonobuoydockerkey                               kubernetes.io/dockerconfigjson        1      16m
```

### daemonsets-6591
```bash
$ kubectl get secrets -n daemonsets-6591
No resources found in daemonsets-6591 namespace.
$ kubectl get serviceaccounts -n daemonsets-6591
No resources found in daemonsets-6591 namespace.
```

**Comment by johnSchnake** [comment]
Ah I see.

So the problem you're running into is that the containers that the e2e process spawns _also_ are in a private registry and the secret isnt available (and also not set on the container to even utilize it if it were)

I'm not certain there is an ideal resolution to this. A little bit of brainstorming though:
 - Ideally you'd be able to put these images into your own public registry and avoid the issue entirely.
 - If you can prepull the images into the cluster then it won't need to pull them again when running in the different namespaces. That is, depending on the tags. I believe if something uses the `:latest` tag then it will still pull as well as if the ImagePullPolicy is Always. See https://kubernetes.io/docs/concepts/containers/images/#pre-pulled-images
 - At the very least, supporting this would require a change to the Kubernetes test framework so that it knows what secrets to move into the namespaces and something like a mutating webhook to update all the containers to set the imagepullsecrets. This would be a pretty sizable change I wouldnt expect it to be implemented any time in the near future.

Unfortunately the framework does just assume that the registries are "public" (not password protected) for most of the images. So you can easily point them to another registry, but not easily change whether it is a public or private one.

**Comment by kobemtl** [comment]
Thank you so much again @johnSchnake . I will try prepull option and see how it works. Otherwise pulling from public is the only way I guess.

**Comment by johnSchnake** [comment]
Good luck. If you hit a bug or just want to talk about the process, let us know. You can file another issue or find us in the k8s slack channel #sonobuoy

**Comment by mdrakiburrahman** [comment]
@johnSchnake - just wanted to share  what's probably a better way to do this - in case it helps unblock others in my shoes. I've read through your analysis here, so I think you'll find the below approach is new (and scalable/easier?):
* https://github.com/vmware-tanzu/sonobuoy/issues/1038
* https://github.com/vmware-tanzu/sonobuoy/issues/424
* https://github.com/vmware-tanzu/sonobuoy/pull/660
* https://github.com/vmware-tanzu/sonobuoy/issues/692

In my use case, we have multiple `imagePullSecrets` that can be used by the test images (we are generating 100s of plugins via `sonobuoy gen plugin > plugin-n.yaml` then executing them in parallel, the worker images are private, unique and different, and each different image can be in a different image registry due to release trains/customizations....complicated, I know).

The YAML injection approach in official docs [here](https://sonobuoy.io/docs/v0.51.0/pullsecrets/) didn't cut it for our use case, even if we scripted out the YAML injection.

Basically, besides injecting the `imagePullSecret` into the Pod, Kubernetes _also_ let's you specify a `ServiceAccount` to have permissions to the Secret. And as long as whatever Pod is using that `ServiceAccount`, it can pull from there: 
* https://dev.to/rizwankh_24/how-to-use-service-accounts-for-kubernetes-imagepullsecrets-3b57

Using this tidbit, I have a "launcher" namespace with a pre-created secret with several (10s) of `imagePullSecrets` (via Kustomize):
```yaml
- name: image-pull-secrets
  namespace: arc-ci-launcher
  options:
    disableNameSuffixHash: true
  files:
  - configs/arc-acr-1-imagePullSecret.json
  - configs/arc-acr-2-imagePullSecret.json
  - ...
```
So Kustomize creates a single humungous secret in `arc-ci-launcher` with all the Docker Configs as K:V pairs with the JSON. This namespace contains a `launcher` Pod, where I have `sonobuoy` CLI ready to trigger `sonobuoy run` - below.

Then, I have this bash function, that simply injects the secret into the Sonobuoy namespace:
```bash
function patch_namespace_imagepull()
{
  if [ $# -lt 3 ];
  then
      echo "Usage: patch_namespace_imagepull <source-namespace> <source-secret> <destination-namespace> <destination-sa>"
      exit 1
  fi

  SOURCE_NS="$1"
  SOURCE_SECRET="$2"
  DEST_NS="$3"
  DEST_SA="$4"

  # Verify secret exists in source namespace
  if ! kubectl get secret -n ${SOURCE_NS} ${SOURCE_SECRET} > /dev/null 2>&1; then
    echo "No \"${SOURCE_SECRET}\" secret found in ${SOURCE_NS}, exiting"
    exit 0
  fi

  SECRETS_JSON=$(kubectl get secret -n ${SOURCE_NS} ${SOURCE_SECRET} -o json | jq -r '.data')
  CONFIGS_BASE64=$(echo "$SECRETS_JSON" | jq -c -r '.[]')

  for CONFIG_BASE64 in ${CONFIGS_BASE64[@]}; do
      # Decode JSON, get the registry url under '.auths', replace '.' with '-'
      CONFIG_JSON=$(echo $CONFIG_BASE64 | base64 -d)
      NEW_DOCKER_SECRET_NAME=$(echo $CONFIG_JSON | jq -r '.auths | keys[]' | sed 's/\./-/g')

      # Create docker config secret in destination namespace
      kubectl create secret generic "${NEW_DOCKER_SECRET_NAME}" \
              -n "${DEST_NS}" \
              --type=kubernetes.io/dockerconfigjson \
              --from-literal=.dockerconfigjson="${CONFIG_JSON}" \
              --dry-run=client -o yaml \
              | kubectl apply -f -

      # Patch the ServiceAccount in the destination namespace. We use jq 
      # because 'kubectl patch' makes it difficult to use nested objects
      kubectl get serviceaccount "${DEST_SA}" -n "${DEST_NS}" -o json \
              | jq ".imagePullSecrets += [{\"name\": \"${NEW_DOCKER_SECRET_NAME}\"}]" \
              | kubectl apply -f -
  done
}
```
Finally, here's the sequence in which I trigger `sonobuoy`, such that the `ServiceAccount` the Pods run under auto inherits all of the above `imagePullSecrets`, making them all work:
```bash
# Pre-create sonobuoy namespace
kubectl create namespace sonobuoy

# Pre-create and patch sonobuoy service account with imagePullSecret env var,
# sourcing the secret from the launcher namespace
kubectl create serviceaccount sonobuoy-serviceaccount --namespace sonobuoy
patch_namespace_imagepull ${LAUNCHER_NAMESPACE} ${IMAGE_PULL_SECRET} sonobuoy sonobuoy-serviceaccount

# Run with existing SA
sonobuoy run --skip-preflight --timeout $timeout --wait \
--service-account-name sonobuoy-serviceaccount \
--existing-service-account \
--plugin $PRE_GENERATED_YAML
```

We wrapped this launcher container up in a bash harness that can scale to N number of plugins, worked beautifully, the only tricky thing is the Launcher Kustomize Secret injection + that bash for-loop patcher `patch_namespace_imagepull` - we found it easier to implement that these Custom Operators though:
* https://github.com/neutryno/imagepullsecret-serviceaccount-patcher
* https://github.com/alexellis/registry-creds
* https://github.com/titansoft-pte-ltd/imagepullsecret-patcher

| P.S. Thanks for maintaining this awesome product :)

**Comment by raosaurabh** [comment]
@johnSchnake Is the issue @kobemtl faced here resolved in latest versions of k8s frameworks? We have a similar use-case and any confirmation is helpful. Thank you.
